design/upgrade.md

```diff
@@ -0,0 +1,165 @@
+If this flag is set, the operator can periodically check Rook's container registry to see if there is a new version tag greater than the version tag currently deployed to the cluster.
+If a new version is found, then the upgrade controller will proceed to update the cluster to the new version in a rolling fashion as described earlier.
+
+It is also possible for the user to be using release channels instead of specific version numbers, e.g. `beta` or `stable`.
+In that case, the operator would be looking for a newer version that has been promoted to the channel that the user has specified.
+
+### `rookctl`
+We should consider implementing a few new commands in `rookctl` that will help the user monitor and verify the upgrade progress and status.
+Some examples for potential new commands would be:
+* `rookctl versions`: This command would return the version of all Rook components in the cluster, so they can see at a glance which components have finished upgrading.
+This is similar to the new [`ceph versions` command](http://ceph.com/community/new-luminous-upgrade-complete/).
+* `rookctl status --upgrade`:  This command would return a summary, retrieved from the upgrade controller, of the most recent completed steps and status of the upgrade that it is currently working on.
+
+### **Migrations and Breaking Changes**
+When a breaking change or a data format change occurs, the upgrade controller will have the ability to automatically perform the necessary migration steps during the upgrade process.
+While migrations are possible, they are certainly not desirable since they require extra upgrade logic to be written and tested, as well as providing new potential paths for failure.
+Going forward, it will be important for the Rook project to increase our discipline regarding the introduction of breaking changes.
+We should be **very** careful about adding any new code that requires a migration during the update process.
+
+### **Kubernetes Built-in Support**
+Kubernetes has some [built-in support for rolling updates](https://kubernetes.io/docs/tasks/run-application/rolling-update-replication-controller/) with the `kubectl rolling-update` command.
+Rook can potentially take advantage of this support for our replication controllers that have multiple stateless pods deployed, such as RGW.
+This support is likely not a good fit for some of the more critical and sensitive components in the cluster, such as monitors, that require careful observation to ensure health is maintained and quorum is reestablished.
+
+If the upgrade controller uses the built-in rolling update support for certain stateless components, it should still verify all cluster health checks before proceeding with the next set of components.
+
+### **Synchronization**
+The upgrade process should be carefully orchestrated in a controlled manner to ensure reliability and success.
```

Comment:
one thing the operator can do is stop watching all CRDs for changes, including cluster, pool, object, file, etc.

Reply:
added a note about not watching CRD events during an upgrade.